### PR TITLE
Propagate anchors from components

### LINKED ISSUE #33: Arabic mark positioning
The file [NotoSansArabic-MM.glyphs](https://github.com/googlei18n/noto-source/blob/master/src/NotoSansArabic-MM.glyphs) from noto-source outputs the following TTF in glyphs:

[NotoSansArabic-Regular.ttf.zip](https://github.com/googlei18n/glyphs2ufo/files/215753/NotoSansArabic-Regular.ttf.zip)

With this TTF it is possible to see that mark U+064B (arabic fathatan) has a positioning rule with base U+FEC0 (arabic letter dad medial form):

```
$ hb-shape NotoSansArabic-Regular.ttf "ﻀً"
[uni064B=0@490,114+0|uniFEC0=0+949]
```

This is despite the fact that in Glyphs' UFO output, glyph U+FEC0 does not have any anchors:

[dad-ar.medi.glif.zip](https://github.com/googlei18n/glyphs2ufo/files/215755/dad-ar.medi.glif.zip)

@schriftgestalt, how is the mark positioning rule determined in this case? Does the output UFO contain the necessary information to produce a correct TTF?

### PR DESCRIPTION
**Pull request by jamesgk**
Should close #33 

cc @behdad

### CODE REVIEW

File: Lib/glyphs2ufo/builder.py [hunk 1]
```diff
@@ -672,6 +673,46 @@ def add_glyph_to_groups(kerning_groups, glyph_data):
...
+        added_here = {}
+
+        # don't propagate anchors for mark glyphs
+        if any(a.name.startswith('_') for a in parent.anchors):
+            continue
+
+        # try to get anchors from base (first) components
+        glyph = parent
+        transformation = Transform()
+        while glyph.components:
```
Comment (dougfelt) [inline Lib/glyphs2ufo/builder.py]: I don't see how this loop terminates.

Comment (jamesgk) [inline Lib/glyphs2ufo/builder.py]: When `glyph` is re-assigned, it's set to the existing `glyph`'s first component (and presumably we won't have an infinite chain of components, i.e. eventually we'll get to a glyph with no components).

Comment (dougfelt) [inline Lib/glyphs2ufo/builder.py]: Oh, duh, you're walking up the chain, glyph keeps changing.  Sorry for being obtuse.

**Comment by jamesgk** [comment]
PTAL. Added one more commit in response to discussion at https://github.com/googlei18n/ufo2ft/pull/34

**Comment by dougfelt** [comment]
LGTM

**Comment by behdad** [comment]
Reading https://glyphsapp.com/tutorials/advanced-diacritics-multiple-anchors looks like underscore has a special meaning, not `_1` or `_2`.  @schriftgestalt can you please clarify where it is stored which alternative anchor a mark is attached to?  I think that's what we need to implement here.

**Comment by schriftgestalt** [comment]
The _1, _2 anchors are usually used on ligatures to generate the liga2mark feature.

**Comment by jamesgk** [comment]
> The _1, _2 anchors are usually used on ligatures to generate the liga2mark feature.

Maybe this is obvious to other people, but I still don't understand how one of these anchors is chosen when attaching a mark. To follow the example in https://github.com/googlei18n/glyphs2ufo/issues/33, we can do:

```
$ hb-shape NotoSansArabic-Regular.ttf "ﻼً"
[uni064B=0@-44,95+0|uniFEFC=0+599]
$ hb-shape NotoSansArabic-Regular.ttf "عقلًا"
[uni064B=2@260,257+0|uniFEFC=2+599|uniFED8=1+495|uniFECB=0+503]
```

U+FEFC (lam_alef-ar.fina) has anchors `top_1 {422, 714}` and `top_2 {118, 552}`, while U+064B (fathatan-ar) has `_top {162, 457}`. So it seems that in the first case the mark is attached to top_2 while in the second it's attached to top_1. Thus this PR is incomplete, but I don't personally know how to fix it.

**Comment by schriftgestalt** [comment]
Are we speaking about the anchors in precomposed glyphs? Or about the mark feature?

the `lam_alef-ar.fina` use a `pos ligature` statement. You can check the feature code in the Temp folder in `~/Library/Application Support/Glyphs/Temp/FontName/features.fea`

**Comment by behdad** [comment]
> Are we speaking about the anchors in precomposed glyphs? Or about the mark feature?

I think the question is, where do you store which of top_x, top_y, top_z a particular top-mark attaches to?

**Comment by jamesgk** [comment]
Well it's now clear to me how to use the _1/2/3 anchors (I wasn't even aware of the `pos ligature` statement... thank you Georg!). I guess it's still an open question how other anchors with underscores in their names are used. Unless, maybe they're not used in mark features at all?

**Comment by schriftgestalt** [comment]
Anchir names with undercores+numbers are not used for mark positioning other than for ligatures. And all anchors that have a character other than a-zA-Z0-9._- should be ignored for the mark features. They are then only used to align components.

**Pull request merged**